Leaflet Map Integration and GUI Improvements

robertlipe opened the pull request:
introduces Leaflet maps into the GPSBabel GUI, laying the groundwork to replace the previous Google Maps implementation for map previews.

Changes and improvements include:
(A ton of stuff that I'll pull into this desription later) 
 - Refactored map preview settings to use a a `BabelData::MapPreviewType` enum
      for consistency and persistence, with migration logic for old settings.
 - Implemented clickable info balloons for waypoint and route point markers on
      the Leaflet map, displaying `name`, `description`, and `comment`.
 - Corrected the `GpxRoutePoint` class to include `description` and `comment`
      members, aligning with `GpxWaypoint` for richer data.
 - Updated browse buttons to use platform-native icons
      (`QStyle::StandardPixmap`).
 - Enhanced `leafletbase.html`, `markerclicker.h`, `leafletmap.h`, and
      `leafletmap.cc` to standardize map marker interactions and popup displays.
 - Overhauled `babeldata.h`, `advdlg.h`, `advdlg.cc`, `preferences.cc`, and
      `mainwindow.cc` to support the new map preview enum.

robertlipe committed:
WIP: Add Leaflet map preview.

This commit introduces a Leaflet-based map preview feature.

- Adds a new map preview dialog based on Leaflet.
- Includes necessary JavaScript and CSS libraries for Leaflet and MarkerCluster.
- Implements the backend logic to generate GeoJSON from GPX data.
- Connects the map preview to the main application.
- Removes trailing whitespace from all staged files.

robertlipe committed:
Fix: Remove duplicate function definition.

robertlipe committed:
Fix: Correct waypoint marker icon anchor in Leaflet map.

robertlipe committed:
Fix: Use standard Leaflet icon to fix marker distortion.

robertlipe committed:
Fix: Rework map visibility logic and disable broken highlighting.

- Rework the JavaScript visibility logic in the Leaflet map to be more robust and prevent synchronization issues between the tree view and the map.
- Disable the waypoint selection highlighting feature, which was causing a JavaScript error after a previous change.

robertlipe committed:
Refactor: Remove Leaflet layer control.

Removes the Leaflet layer control from the map UI and all associated C++ code. This simplifies the UI and eliminates a source of synchronization bugs. The Qt tree view is now the single source of truth for layer visibility.

robertlipe committed:
Fix: Robustly synchronize checkbox state in map dialog.

- Replaces signal blocking with a re-entrancy guard to ensure the UI updates correctly.
- When a parent checkbox is toggled, its state is now correctly propagated to all its children.
- When child checkboxes are toggled, the parent state is now correctly updated to reflect whether all, some, or none of the children are checked.

robertlipe committed:
Fix: Correct JavaScript calls for route visibility.

- Updates the C++ code to call the correct JavaScript functions for setting route visibility.
- This fixes a JavaScript error caused by referencing a non-existent variable from a previous implementation.

robertlipe committed:
Fix: Correct checkbox sync and route drawing in map dialog.

- Fixes a bug that prevented routes from being displayed on the map.
- Implements a re-entrancy guard to correctly handle checkbox state synchronization between parent and child items in the tree view.
- Removes a duplicate member variable declaration.

robertlipe committed:
Fix: Implement route serialization in GeoJSON writer.

- Adds logic to the GeoJSON writer to correctly serialize routes.
- Routes are now represented as a LineString feature with an embedded property containing an array of the individual route points.
- The Leaflet map is updated to parse this new structure and render routes as a styled path with clickable markers for each point.
- Fixes duplicate function declarations in geojson.h.

robertlipe committed:
Docs: Document rationale for internal GeoJSON generation.

- Adds internal documentation to mainwindow.cc explaining why generateGeoJsonWithIndices is used for the GUI map preview instead of the CLI geojson writer.
- Clarifies the need for GUI-specific properties like originalIndex for interactive features.

robertlipe committed:
Fix: Embed route points in GeoJSON properties for GUI map preview.

- Modifies MainWindow::generateGeoJsonWithIndices to correctly embed route points (lat, lng, name) as a 'routepoints' array within the properties of the route's LineString feature.
- This ensures that individual route points can be displayed as clickable markers on the Leaflet map preview.

robertlipe committed:
Feat: Implement click handling and visual consistency for routes and tracks.

- Route point markers are now red for visual consistency with route paths.
- Implements click handling for individual route points, allowing them to be selected and highlighted in the tree view.
- Implements click handling for tracks, allowing them to be selected and highlighted in the tree view.
- Fixes various signal/slot connection issues and missing declarations.

robertlipe committed:
Fix: JavaScript syntax error in leafletbase.html.

- Adds a missing closing brace for the 'if (coords.length > 0)' block within the track feature processing.
- This resolves the 'Uncaught SyntaxError: Unexpected token ')' error.

robertlipe committed:
Fix: Make route linestrings clickable.

- Adds an attachHandler to the route linestring (pathLayer) to enable click events and highlight the corresponding route in the tree view.

robertlipe committed:
Debug: Add console.log statements for click event debugging.

robertlipe committed:
Add a transparent track to make tracks clickable

robertlipe committed:
Reconcile the aborted multi-checkbox settings for Leaflet vs. GMaps setting. Fix issue when both are selected in preferences object. Add runtime upgrade path for settings.

robertlipe committed:
Add balloons to wpt and rpt

robertlipe committed:
Don't remediate preferences if we have new ones. Otherwise, we can't ever save Leaflet mode

robertlipe committed:
Replace windows 95 icons with native ones. Not sure I love this, though...

robertlipe committed:
Turn down JS debugging

robertlipe committed:
feat: Reintroduce marker clustering for waypoints in Leaflet map and fix maxZoom issue

robertlipe committed:
feat: Reintroduce marker clustering, fix icon scaling, and simplify waypoint interaction in Leaflet map

robertlipe commented on gui/leafletmap.cc:
```diff
@@ -0,0 +1,339 @@
+void
+LeafletMap::resetBounds()
+{
+  evaluateJS(QStringList{
```
Why a QStringList here?

robertlipe commented on gui/mainwindow.cc:
```diff
@@ -229,6 +248,10 @@ MainWindow::MainWindow(QWidget* parent): QMainWindow(parent)
 
   //--- Restore from registry
   restoreSettings();
+// RJL
```
Remove these

robertlipe commented on gui/mainwindow.cc:
```diff
@@ -1176,7 +1228,7 @@ void MainWindow::preferencesActionX()
   preferences.exec();
 
-  // The user may have changed the list of displayed formats, and/or
+  // The user may have changed the list of of displayed formats, and/or
```
of of

robertlipe commented on gui/webenginepage.cc:
```diff
@@ -0,0 +1,44 @@
+// -*- C++ -*-
```
copyright

robertlipe commented on gui/webenginepage.h:
```diff
@@ -0,0 +1,22 @@
+// -*- C++ -*-
```
copyright

robertlipe commented on gui/leafletdlg.cc:
```diff
@@ -0,0 +1,559 @@
+// -*- C++ -*
```
copyright

robertlipe commented on gui/leafletmap.cc:
```diff
@@ -0,0 +1,339 @@
+// -*- C++ -*
```
copyright

robertlipe commented on gui/leafletdlg.h:
```diff
@@ -0,0 +1,67 @@
+// -*- C++ -*-
```
copyright

tsteven4 commented:
in leafletmap.cc I think you can set the user agent with something like this->page()->profile()->setHttpUserAgent("my user agent id")

robertlipe committed:
Whack QNetworkAccessManager

robertlipe commented:
Probably.  I'm not opposed to it and will probably fold this in.  My read
was that Leaflet, not us, is the "customer" here. They control the caching,
fetching, etc.  We're no more responsible than, say, Chrome is for
loading leaflet. Is the correct User-agent Qt or Blink (open source Chrome
that's used in Qt) or Leaflet or us?

There's some ambiguity <https://github.com/Leaflet/Leaflet/pull/9883> here,
but it seems they hard block anything that doesn't meet their
requirements.
https://community.openstreetmap.org/t/access-blocked-leaflet-tiles-just-me/135657
(Hey, I know one of those posters... He's a keeper of GSAK, which uses
GPSBabel after the creator suffered a stroke several years ago.) Leaflet
doesn't mention it in their QSG
<https://leafletjs.com/examples/quick-start/>.

Re: choice of maps - There's an embarrassment of riches (Mapbox, Stadia,
Thunderforest, Lima, Bing, Fleaflet
<https://docs.fleaflet.dev/tile-servers/using-openstreetmap-direct>, OSM
and more OSM) of tile servers that we probably COULD use and I think we'd
be under the free limits of all of them. (I'm not saying that all of these
ARE compatible; it's just q quick eample list).

One of my earliest explorations into Digital carto was requesting the CD
set of the US Tiger network and experimenting with what would now be a tile
service. In 2000, the tools to do this were totally terrible, but that was
also the days when ordering a half dozen CDs would deliver the bits faster
than my ISDN connection would have. Slinging around a couple hundred GB is
a lot easier these days. I started to go down a rabbit hole of enabling a
bunch and then went back to the tagline that I had to keep using on myself:
"We're not a mapping app."

I don't plan to engage in behaviour I wouldn't sign my name to, but I'm
genuinely unsure if it's ours to sign. Anyone engaging in something sketchy
can find WAY easier ways of accomplishing it than driving the maps through
GPSBabel or commenting out that line above.  I'd rather have the discussion
of "it wasn't us" than even the appearance that we're trying to dodge
anything. We're all just trying to provide things for others for free and
NOT get ripped off in the process.

Sidebar: I tried the agentic UI in Copilot on GitHub for the first time
with a glorified version of "go do the slugwork of addressign the reviewer
comments in this PR".  It had high-impact things like replacing "to to" and
bogus indention, though I DID coerce it to refactor gui/gpx.h to derive
from a base class. Two hours ago it told me it was working on it.  I'm not
sure if that's any better than telling Cursor, VS Code, Windsurf, CLion, or
any of the other editors to deal with it. Like half the programming world,
I'm still trying to figure out when AI coding helpers are actually helpful.
Could I have changed "to to" faster than this? Certainly. But outsourcing
slugwork is certainly appealing at some level. I don't have interns any
more. :-)

Still, I'll roll something like this in. Thanx.



On Thu, Oct 23, 2025 at 1:49 PM codacy-production[bot] <
***@***.***> wrote:

> *codacy-production[bot]* left a comment (GPSBabel/gpsbabel#1498)
> <https://github.com/GPSBabel/gpsbabel/pull/1498#issuecomment-3438608963>
> Coverage summary from Codacy See diff coverage on Codacy
> <https://app.codacy.com/gh/GPSBabel/gpsbabel/pull-requests/1498/coverage?utm_source=github.com&utm_medium=coverageSummary&utm_campaign=coverageSummaryPullRequest>
> Coverage variation Diff coverage
> ✅ *-0.10%* ✅ *12.00%* Coverage variation details
> Coverable lines Covered lines Coverage
> Common ancestor commit (40e339e
> <https://github.com/GPSBabel/gpsbabel/commit/40e339e4b3b6fbb3f8584e1c8f8b1fa7d1fb6135>
> ) 27929 17545 62.82%
> Head commit (dd79fe7
> <https://github.com/GPSBabel/gpsbabel/commit/dd79fe7bbd0302b452edae4812caff6e4bb9576f>
> ) 27975 (+46) 17547 (+2) 62.72% (*-0.10%*)
>
> *Coverage variation* is the difference between the coverage for the head
> and common ancestor commits of the pull request branch: <coverage of head
> commit> - <coverage of common ancestor commit>
> Diff coverage details
> Coverable lines Covered lines Diff coverage
> Pull request (#1498 <https://github.com/GPSBabel/gpsbabel/pull/1498>) 50 6
> *12.00%*
>
> *Diff coverage* is the percentage of lines that are covered by tests out
> of the coverable lines that the pull request added or modified: <covered
> lines added or modified>/<coverable lines added or modified> * 100%
> See your quality gate settings
> <https://app.codacy.com/gh/GPSBabel/gpsbabel/settings/quality?utm_source=github.com&utm_medium=coverageSummary&utm_campaign=coverageSummaryGates>
>     Change summary preferences
> <https://app.codacy.com/gh/GPSBabel/gpsbabel/settings/integrations?utm_source=github.com&utm_medium=coverageSummary&utm_campaign=coverageSummarySettings>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GPSBabel/gpsbabel/pull/1498#issuecomment-3438608963>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCSD342UTOXRQSLYP5ME5T3ZEPNNAVCNFSM6AAAAACJ4IB4FOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMZYGYYDQOJWGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tsteven4 commented:
> 2\. gmaps is broken,  e.g. with bounds_test.gpx:
>     js: InvalidValueError: setPosition: not a LatLng or LatLngLiteral: in property lat: not a number
>     ...
>     js: InvalidValueError: setPosition: not a LatLng or LatLngLiteral: in property lat: not a number
>     js: Uncaught InvalidValueError: not a LatLng or LatLngLiteral: not an Object

This, and a crash with google maps, is fixed with the following patch.  This works for both google maps and leaflet.
[gmap.patch](https://github.com/user-attachments/files/23109792/gmap.patch)

robertlipe committed:
Address a ton of review notes - mostly mine.

Still WIP

robertlipe committed:
Reindent mainwin new function

robertlipe committed:
Apply tsteven4's fix for Google Maps crash.

From https://github.com/GPSBabel/gpsbabel/pull/1498#issuecomment-3439626769

robertlipe committed:
Merge branch 'master' into leaflet

robertlipe committed:
Merge branch 'leaflet' of https://github.com/GPSBabel/gpsbabel into leaflet

robertlipe committed:
Integrate review feedback on Leaflet.

Don't disconnec the signals to avoid reentry. Just block the signals.
Since itemChangedX does the deed, make it responsible for a visual
refresh on changes, letting icons follow check status. Simplified
showHideAll()

showOnlyThis left unselected items on the map when it disconnected.
Don't do that. Block the signals, toggle ALL the checkmarks, hide
everything in Javascript-land, then re-enable the single item, unblock.
Vastly reduces bridge transactions.

robertlipe committed:
Let leaflet inherit injected babelData_.debugLevel.

Ordering fixes.
Remove debug code that clobbered output file.

robertlipe committed:
Merge branch 'master' into leaflet

robertlipe committed:
Merge branch 'master' into leaflet
